### ISSUE TITLE
Confusing interface to mesh points

### ISSUE THREAD

**Issue by mwallerb** [ALPSCore/ALPSCore#304]
We have the distinction between, e.g., `alps::gf::matsubara_mesh`, which is conceptually a set of points, and a `alps::gf::matsubara_index`, which is a typed index into that set of points.

But while indexing a Green's function using `operator()` with an index yields the appropriate value, this is not true for the mesh: `alps::gf::matsubara_mesh::operator()` instead returns back the integer value behind the index rather than the value of the corresponding Matsubara frequency.
```
     alps::gf::matsubara_mesh iw_mesh(20, 10.);
     alps::gf::matsubara_index iw_index(3);
     std::cout << iw_mesh(iw_index);   // prints 3
     std::cout << iw_mesh.points()[iw_index()]   // Workaround: prints the value
```

This is IMHO clearly an interface bug.

**Comment by kleinhenz** [comment]
I guess the intention is to have the call operator of the `mesh` object map from the "physical" value of the index which may be negative (e.g. for negative matsubara frequencies) to the actual data index? IMHO this makes sense but should probably not use the call operator since as noted by @mwallerb the call operator is used for indexing the actual `gf` objects. Would it make sense to define a member function `int dataidx(index_type idx)` to do the conversion and then defining the call operator as `double operator()(index_type idx) const {return _points()[dataidx(idx)];}`?

**Comment by galexv** [comment]
Yes, this is a sound option --- and i am working on it right now. :)

**Comment by galexv** [comment]
@iskakoff Sergey, if you are anyway fixing GFs, are you able to address this issue also? (Otherwise it can be postponed till the new GFs are stabilized a bit)

**Comment by egull** [comment]
Sergey, I assigned this to you. Can you check if you want to do anything with it, or if you're happy with what we have?

**Comment by iskakoff** [comment]
@egull I am not happy, but it can break backward compatibility

**Comment by egull** [comment]
@mwallerb do you have an idea of how to fix it without breaking the interface? We can't really have 20+ codes do a rewrite...

**Comment by mwallerb** [comment]
Stage 1 solution: add docs.
Stage 2 solution: add docs + provide helper for this:

    iw_mesh.points()[iw_index()]

**Comment by iskakoff** [comment]
@mwallerb Stage 3: Profit!
